### PR TITLE
planner: refactor some code of cross-db binding

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #51347

Problem Summary: planner: refactor some code of cross-db binding

### What changed and how does it work?

planner: refactor some code of cross-db binding

NO logical change, just refactor.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/bindinfo/binding_cache.go [hunk 1]
```diff
@@ -48,6 +48,101 @@ var GetBindingReturnNilAlways = stringutil.StringerStr("getBindingReturnNilAlway
...
+	SQLDigest2NoDBDigest(sqlDigest string) string
+
+	// Copy copies this digestBiMap.
+	Copy() digestBiMap
+}
+
+type digestBiMapImpl struct {
+	mu                   sync.RWMutex
+	noDBDigest2SQLDigest map[string][]string // noDBDigest --> sqlDigests
+	sqlDigest2noDBDigest map[string]string   // sqlDigest --> noDBDigest
```
Comment (hawkingrei) [inline pkg/bindinfo/binding_cache.go:75]: sqlDigest2noDBDigest's value can use the ```[unique](https://pkg.go.dev/unique)``` to reduce memory

**Review by hawkingrei — APPROVED** [review]
Except ```unique```, I LGTM all the others

File: pkg/bindinfo/binding_cache.go (hunk 1 above)
Comment (qw4990) [reply pkg/bindinfo/binding_cache.go:75]: OK I'll update here later, I have a few more refactoring PRs based on this, after merging them all I'll update here.

Comment (AilinKid) [reply pkg/bindinfo/binding_cache.go:75]: unique handle seems interesting

File: pkg/bindinfo/binding_cache.go [hunk 2]
```diff
@@ -48,6 +48,101 @@ var GetBindingReturnNilAlways = stringutil.StringerStr("getBindingReturnNilAlway
...
+}
+
+func newDigestBiMap() digestBiMap {
+	return &digestBiMapImpl{
+		noDBDigest2SQLDigest: make(map[string][]string),
+		sqlDigest2noDBDigest: make(map[string]string),
+	}
+}
+
+func (b *digestBiMapImpl) Add(noDBDigest, sqlDigest string) {
```
Comment (AilinKid) [inline pkg/bindinfo/binding_cache.go:85]: exported func should have some comments

File: pkg/bindinfo/binding_cache.go [hunk 3]
```diff
@@ -48,6 +48,101 @@ var GetBindingReturnNilAlways = stringutil.StringerStr("getBindingReturnNilAlway
...
+	return b.noDBDigest2SQLDigest[noDBDigest]
+}
+
+func (b *digestBiMapImpl) SQLDigest2NoDBDigest(sqlDigest string) string {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.sqlDigest2noDBDigest[sqlDigest]
+}
+
+func (b *digestBiMapImpl) Copy() digestBiMap {
```
Comment (AilinKid) [inline pkg/bindinfo/binding_cache.go:127]: what the case of cp a map here？

**Review by AilinKid — COMMENTED** [review]
rest LGTM

File: pkg/bindinfo/binding_cache.go (hunk 3 above)
Comment (qw4990) [reply pkg/bindinfo/binding_cache.go:127]: When we load new bindings, we'll copy the old binding cache and then insert new records into it and then replace it with the old one.

File: pkg/bindinfo/binding_cache.go (hunk 2 above)
Comment (qw4990) [reply pkg/bindinfo/binding_cache.go:88]: added

**Review by AilinKid — APPROVED** [review]

**Pull request merged**